File: frontend/apps/web-antd/src/api/admin/operation-log.ts
/**
 * 操作日志 API
 * 对接后端 /admin/operation-logs/* 接口
 */
import type { ApiRequestOptions } from '#/utils/request';

import { requestClient } from '#/utils/request';

// ============================================================
// 类型定义
// ============================================================

/** 操作日志列表查询参数 */
export type OperationLogListParams = Record<string, unknown>;

/** 操作日志信息（后端原始格式 snake_case） */
export interface OperationLogInfoRaw {
  id: number;
  tenant_id: null | number;
  user_type: string;
  username: null | string;
  nickname?: null | string;
  module: string;
  module_label?: string;
  action: string;
  action_label?: string;
  method: string;
  path: string;
  query_params: null | Record<string, unknown>;
  request_body: null | Record<string, unknown>;
  status_code: number;
  response_code: number;
  ip: string;
  user_agent: null | string;
  duration_ms: number;
  created_at: string;
}

/** 操作日志信息（前端格式 camelCase） */
export interface OperationLogInfo {
  id: number;
  tenantId: null | number;
  userType: string;
  username: null | string;
  nickname?: null | string;
  module: string;
  moduleLabel?: string;
  action: string;
  actionLabel?: string;
  method: string;
  path: string;
  queryParams: null | Record<string, unknown>;
  requestBody: null | Record<string, unknown>;
  statusCode: number;
  responseCode: number;
  ip: string;
  userAgent: null | string;
  durationMs: number;
  createdAt: string;
}

/** 分页列表响应 */
export interface OperationLogListResponse {
  items: OperationLogInfo[];
  total: number;
  page: number;
  page_size: number;
}

// ============================================================
// 转换函数
// ============================================================

/** 将后端 snake_case 转换为前端 camelCase */
function transformOperationLogInfo(raw: OperationLogInfoRaw): OperationLogInfo {
  return {
    id: raw.id,
    tenantId: raw.tenant_id,
    userType: raw.user_type,
    username: raw.username,
    nickname: raw.nickname,
    module: raw.module,
    moduleLabel: raw.module_label,
    action: raw.action,
    actionLabel: raw.action_label,
    method: raw.method,
    path: raw.path,
    queryParams: raw.query_params,
    requestBody: raw.request_body,
    statusCode: raw.status_code,
    responseCode: raw.response_code,
    ip: raw.ip,
    userAgent: raw.user_agent,
    durationMs: raw.duration_ms,
    createdAt: raw.created_at,
  };
}

// ============================================================
// API 接口
// ============================================================

const API_PREFIX = '/admin/operation-logs';

/**
 * 获取操作日志列表
 * GET /admin/operation-logs
 */
export async function getOperationLogListApi(
  params?: OperationLogListParams,
  options?: ApiRequestOptions,
): Promise<OperationLogListResponse> {
  const response = await requestClient.get<{
    items: OperationLogInfoRaw[];
    page: number;
    page_size: number;
    total: number;
  }>(API_PREFIX, { params, ...options });

  return {
    items: response.items.map((item) => transformOperationLogInfo(item)),
    total: response.total,
    page: response.page,
    page_size: response.page_size,
  };
}

/**
 * 获取操作日志详情
 * GET /admin/operation-logs/{id}
 */
export async function getOperationLogDetailApi(
  id: number,
  options?: ApiRequestOptions,
): Promise<OperationLogInfo> {
  const raw = await requestClient.get<OperationLogInfoRaw>(
    `${API_PREFIX}/${id}`,
    options,
  );
  return transformOperationLogInfo(raw);
}

/**
 * 删除操作日志（支持批量）
 * DELETE /admin/operation-logs
 */
export async function deleteOperationLogsApi(
  ids: number[],
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.delete(API_PREFIX, {
    data: { ids },
    ...options,
  });
}

File: backend/app/models/system/operation_log.py
"""
操作日志模型 / Operation Log Model

记录系统中所有 API 调用的审计日志
Records audit logs for all API calls in the system.
"""


from sqlalchemy import JSON, ForeignKey, Index, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.base_model import BaseModel


class OperationLog(BaseModel):
    """
    操作日志模型 / Operation log model.

    记录所有 API 调用的审计日志，支持：
    - 平台端操作（tenant_id 为空）
    - 企业端操作（tenant_id 不为空）
    """

    __tablename__ = "operation_logs"

    __ai_policy__ = {
        "label": "操作日志",
        "keywords": ["日志", "audit", "审计"],
        "allow_read": True,
        "blocked_columns": ["request_body", "query_params", "ip", "user_agent"],
    }

    # 表级索引 / Table-level indexes
    __table_args__ = (
        # 复合索引：企业 + 时间，用于企业日志查询 / tenant + time for tenant log queries
        Index("ix_operation_logs_tenant_created", "tenant_id", "created_at"),
        # 复合索引：用户类型 + 用户ID，用于用户操作追溯 / user type + id trace
        Index("ix_operation_logs_user", "user_type", "user_id"),
        # 复合索引：模块 + 操作，用于操作统计 / module + action stats
        Index("ix_operation_logs_module_action", "module", "action"),
    )

    # 可过滤字段声明 / Declared filterable fields
    __filterable__ = {
        "id": "id",
        "trace_id": "trace_id",
        "tenant_id": "tenant_id",
        "user_type": "user_type",
        "user_id": "user_id",
        "username": "username",
        "module": "module",
        "action": "action",
        "resource": "resource",
        "method": "method",
        "path": "path",
        "status_code": "status_code",
        "response_code": "response_code",
        "ip": "ip",
        "created_at": "created_at",
    }

    __sortable__ = ["id", "username", "module", "action", "method", "response_code", "created_at"]

    # ==================== 用户信息 ==================== / User identity

    # 企业ID（平台操作时为空） / Tenant id (null for platform ops)
    tenant_id: Mapped[int | None] = mapped_column(
        Integer,
        ForeignKey("tenants.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="企业ID（平台操作为空）"
    )

    # 用户类型: admin / tenant_admin / tenant_user
    user_type: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        index=True,
        comment="用户类型"
    )

    # 用户ID
    user_id: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
        comment="用户ID"
    )

    # 用户名（冗余存储，便于查询和展示） / Username denormalized
    username: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
        comment="用户名"
    )

    # 用户昵称（冗余存储，便于展示） / Nickname denormalized
    nickname: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        comment="用户昵称"
    )

    # ==================== 操作信息 ==================== / Operation details

    # 业务模块: auth / permission / role / admin_user / tenant / config / plan / ...
    module: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
        index=True,
        comment="业务模块"
    )

    # 操作类型: create / update / delete / query / login / logout / export / import
    action: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
        index=True,
        comment="操作类型"
    )

    # 资源标识（权限代码）: admin_user:create / role:update 等
    resource: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
        comment="资源标识（权限代码）"
    )

    # ==================== 追踪信息 ==================== / Tracing

    # 请求追踪 ID（用于关联日志、审计、Celery 任务）
    trace_id: Mapped[str | None] = mapped_column(
        String(64),
        nullable=True,
        index=True,
        comment="请求追踪 ID / Trace ID for request correlation",
    )

    # ==================== 请求信息 ==================== / Request

    # HTTP 方法: GET / POST / PUT / DELETE / PATCH
    method: Mapped[str] = mapped_column(
        String(10),
        nullable=False,
        comment="HTTP 方法"
    )

    # 请求路径 / Request path
    path: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        index=True,
        comment="请求路径"
    )

    # 查询参数（JSON 格式）
    query_params: Mapped[dict | None] = mapped_column(
        JSON,
        nullable=True,
        comment="查询参数"
    )

    # 请求体摘要（JSON 格式，已脱敏）
    request_body: Mapped[dict | None] = mapped_column(
        JSON,
        nullable=True,
        comment="请求体摘要（已脱敏）"
    )

    # ==================== 响应信息 ==================== / Response

    # HTTP 响应状态码
    status_code: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
        comment="HTTP 响应状态码"
    )

    # 业务响应码 / App response code
    response_code: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
        index=True,
        comment="业务响应码"
    )

    # 响应消息 / Response message
    response_message: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
        comment="响应消息"
    )

    # ==================== 客户端信息 ==================== / Client

    # 客户端 IP
    ip: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
        index=True,
        comment="客户端 IP"
    )

    # User-Agent / 浏览器 UA 字符串
    user_agent: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
        comment="User-Agent"
    )

    # ==================== 性能指标 ==================== / Performance

    # 请求耗时（毫秒） / Request duration (ms)
    duration_ms: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
        comment="请求耗时（毫秒）"
    )

    def __repr__(self) -> str:
        return (
            f"<OperationLog(id={self.id}, user={self.username}, "
            f"action={self.action}, path={self.path})>"
        )


__all__ = ["OperationLog"]
